### ISSUE TITLE
Unicode literals interfere with config type checking

### ISSUE THREAD

**Issue by lehmannro** [sphinx-doc/sphinx#2326]
The configuration type checker will complain if using `unicode` in place of `str`.  This is exacerbated by using `from  __future__ import unicode_literals`, which will turn all strings into `unicode`.

Just replacing all explicit `str` type hints with `basestring` (or the `six` equivalent of it) won't do enough:  The inferred type hints would still be `str` only.

The best course of action is to treat `unicode` and `str` alike in Python 2 (optionally: only of `compile(...confpy...).co_flags & __future__.unicode_literals.compiler_flag`.)

**Comment by timgraham** [comment]
Milestone should be 1.4 -- this doesn't affect the 1.3.x series.

**Comment by tk0miya** [comment]
> The inferred type hints would still be str only.

`check_type()` tries to compare user inputs with default value using common base classes.
It seems this check is ok to str and unicode.

``` python
>>> type('')
<type 'str'>
>>> type('').__bases__
(<type 'basestring'>,)
>>> type(u'').__bases__
(<type 'basestring'>,)
>>> common = (set(type('').__bases__ + (type(''),)) & set(type(u'').__bases__))
>>> common
set([<type 'basestring'>])
```

So we have to take care of explicit type hinting.
I think it is simple to add both str and unicode to type hints.

```
string_types = [six.text_types]  # => [unicode] (py2) or [str] (py3)
if PY2:
    string_types.append(six.binary_types)  # => [str, unicode] (py2)

...

today_fmt = (None, 'env', string_types),
```

**Comment by tk0miya** [comment]
I created new pull request #2342 to fix this.
could you review it please?

**Comment by tk0miya** [comment]
I fixed this at #2342.

Thanks,
